core/cas-server-core-web/src/main/java/org/apereo/cas/web/support/WebUtils.java

```diff
@@ -390,8 +390,19 @@ public static void putRegisteredService(final RequestContext context, final Regi
     public static Credential getCredential(final RequestContext context) {
         final Credential cFromRequest = (Credential) context.getRequestScope().get(PARAMETER_CREDENTIAL);
         final Credential cFromFlow = (Credential) context.getFlowScope().get(PARAMETER_CREDENTIAL);
+        final Credential cFromConversation = (Credential) context.getConversationScope().get(PARAMETER_CREDENTIAL);
 
-        Credential credential = cFromRequest != null ? cFromRequest : cFromFlow;
+        Credential credential = cFromRequest;
+        if (credential == null || StringUtils.isBlank(credential.getId())) {
+            credential = cFromFlow;
+        }
+        if (credential == null || StringUtils.isBlank(credential.getId())) {
+            credential = cFromConversation;
+            if (credential != null && !StringUtils.isBlank(credential.getId())) {
+                //aup and some other modules look only in flow scope via expressions, push down if needed
+                context.getFlowScope().put(PARAMETER_CREDENTIAL, credential);
```

Comment:
Use `WebUtils` please and the relevant method that puts credentials into the flow.

Reply:
This is WebUtils :)

Reply:
Of course it is! :) Sorry.